tests/Examples/UnionExample.hs

```diff
@@ -1,42 +1,47 @@
 {-# LANGUAGE DataKinds #-}
 module Examples.UnionExample  where
 
--- TODO: union code is totally wrong because I misunderstood the
--- spec. The server decides the return type and we'll probably need an
--- open sum type for returning.
-
-{-
-
-import Protolude hiding (Enum)
-import qualified GraphQL.Internal.AST as AST
-import Data.Attoparsec.Text (parseOnly, endOfInput)
-import GraphQL.Internal.Parser (queryDocument)
-
-
+import Protolude hiding (Enum, U1)
+import qualified GraphQL.Internal.Validation as Validation
 import GraphQL.API
-import GraphQL.Server
+import GraphQL (compileQuery, getOperation)
+import GraphQL.Resolver
 import GraphQL.Value (Value)
 
-type O1 = Object "O1" '[] '[Field "o1" Text]
-type O2 = Object "O2" '[] '[Field "o2" Text]
+-- Slightly reduced example from the spec
+type MiniCat = Object "MiniCat" '[] '[Field "name" Text, Field "meowVolume" Int32]
+type MiniDog = Object "MiniDog" '[] '[Field "barkVolume" Int32]
+
+type CatOrDog = Union "CatOrDog" '[MiniCat, MiniDog]
+type CatOrDogList = List (Union "CatOrDog" '[MiniCat, MiniDog])
+
+miniCat :: Text -> Handler IO MiniCat
+miniCat name = pure (pure name :<> pure 32)
 
-type T = Union "U" '[O1, O2]
+miniDog :: Handler IO MiniDog
+miniDog = pure (pure 100)
 
-o1 :: Handler IO O1
-o1 = pure (pure "hello from O1")
+catOrDog :: Handler IO CatOrDog
+catOrDog = do
+  name <- pure "Hello" -- we can do nomadic actions
+  unionValue @MiniCat (miniCat name)
 
-o2 :: Handler IO O2
-o2 = pure (pure "hello from O2")
+catOrDogList :: Handler IO CatOrDogList
+catOrDogList =
+  [ unionValue @MiniCat (miniCat "Felix")
+  , unionValue @MiniCat (miniCat "Mini")
+  , unionValue @MiniDog miniDog
+  ]
 
-tHandler :: Handler IO T
-tHandler = o1 :<|> o2
+exampleQuery :: IO (Result Value)
+exampleQuery = buildResolver @IO @CatOrDog catOrDog (query "{ ... on MiniCat { name meowVolume } ... on MiniDog { barkVolume } }")
 
-exampleQuery :: IO Value
-exampleQuery = buildResolver @IO @T tHandler (query "{ ... on O1 { o1 } ... on O2 { o2 } }")
+-- unionValue can also be used in a list:
+exampleListQuery :: IO (Result Value)
+exampleListQuery = buildResolver @IO @CatOrDogList catOrDogList  (query "{ ... on MiniCat { name meowVolume } ... on MiniDog { barkVolume } }")
```

Comment:
It's pretty long and looks like this: 

```
Result [] (ValueObject (Object {objectFields = [ObjectField (Name {getNameText = "name"}) (ValueString (String "Hello")),ObjectField (Name {getNameText = "meowVolume"}) (ValueInt 32)]}))
```

and we'll have to keep it up to date by hand - not sure that's a good idea?